### ISSUE TITLE
Try to reduce overhead of abstraction in Utf-8 integer formatter APIs

### ISSUE THREAD

**Issue by ahsonkhan** [dotnet/corefx#25425]
Currently, the overhead to calling the Utf-8 integer formatter APIs is high for the simple/default case.

Let's say we just want to write the integer starting from the end of the buffer, and we provide a StandardFormat 'R' to support that.

The function calls look like this:
TryFormat(int value, ..., 'R') => TryFormatInt64(value, 0xffffffff, buffer, out bytesWritten, format) => if check + switch statement => TryFormatInt64R
https://github.com/dotnet/corefx/blob/master/src/System.Memory/src/System/Buffers/Text/Utf8Formatter/Utf8Formatter.Integer.Signed.cs#L18

Even if TryFormatInt64R is essentially a no-op, the overhead is higher than implementing your own (relatively naiive) method and calling it directly.
```C#
private static bool TryFormatInt64R(long value, Span<byte> buffer, out int bytesWritten)
{
    bytesWritten = 0;
    return false;
}
```

```C#
// Note: Not tested, just an example
public static bool MyReverseFormat(uint value, Span<byte> buffer, out int bytesWritten)
{
    int digits = 0;
    bytesWritten = 0;
    do
    {
        int index = buffer.Length - 1 - digits;
        if (index < 0) 
        {
            buffer.Clear();
            return false;
        }
        buffer[index] = (byte)('0' + value % 10);
        value /= 10;
        digits++;
    } while (value > 0);

    bytesWritten = digits;
    return true;
}
```

**Overhead of calling no-op TryFormat:**
![image](https://user-images.githubusercontent.com/6527137/33107176-e63797b6-ceea-11e7-8f32-3d45b416ac53.png)

**We never overcome the overhead, even though the per loop iteration performance of our algorithm is better:**
![image](https://user-images.githubusercontent.com/6527137/33107144-ba032dae-ceea-11e7-9612-2ddeb87b9cf1.png)

![image](https://user-images.githubusercontent.com/6527137/33107156-d08d0aea-ceea-11e7-8bc5-7019bf4176cb.png)

cc @KrzysztofCwalina, @AtsushiKan, @jkotas

**Comment by jkotas** [comment]
What is StandardFormat 'R' ? I do not see it anywhere in https://github.com/dotnet/corefx/blob/d9924a578be5dab9237eaa2d324ecd66488eab96/src/System.Memory/src/System/Buffers/Text/Utf8Formatter/Utf8Formatter.Integer.Signed.cs#L18

Well, as long as you have the StandardFormat or any kind of format thing anywhere on the path to do plain vanilla integer formatting, it will be always significantly slower than rolling your own plain vanilla formatter for small integers.

**Comment by ahsonkhan** [comment]
> What is StandardFormat 'R' ? I do not see it anywhere in

That is a format I was considering adding support for (to indicate write at the end of the buffer, instead of at the start). It isn't there yet. I was using it as an example.

> it will be always significantly slower than rolling your own plain vanilla formatter for small integers.

It is slower for any integer, small or large, but I see your point.

**Comment by jkotas** [comment]
> to indicate write at the end of the buffer

I do not think we should be adding hard-to-use micro-optimized APIs like this. The extra cycles saved by this hack are not that much.

> It is slower for any integer, small or large, but I see your point.

Right. It hurts the most for small integers. It gets amortized to some degree for large integers.

**Comment by jkotas** [comment]
BTW: The current implementation will be also significantly slower compared to hand-rolled implementation for 32-bit integers on 32-bit platforms because of it really formats 64-bit number underneath. The 64-bit math is multiple times slower compared to 32-bit math on 32-bit platforms.

**Comment by KrzysztofCwalina** [comment]
>I do not think we should be adding hard-to-use micro-optimized APIs like this. The extra cycles saved by this hack are not that much.

There are protocols that use length prefix formats. Formatting backwards in front of a buffer is super useful in such situations. Why do you call it a hack? It's a bit more tricky to use it, but not that diffuclt:

Span<byte> buffer = ...
int written = WritePayload(buffer.Slice(20));
Utf8Formatter.TryFormat(buffer.Slice(0, 20), written, 'R');

**Comment by jkotas** [comment]
In your example, I think it would be fine to format into scratch buffer and then copy the Span over where it needs to be. To make that work well, we need to have good fast basic building blocks:

- Format integer: We do not have a fast way to UTF8 integer formatting: All APIs force you to pay the StandardFormat penalty.
- Copy span: Span copy has significant overhead for small spans (https://github.com/dotnet/coreclr/issues/15076)

If we really want to have micro-optimized APIs like that, I do not think it should be a format specifier. It should be a different set of APIs, something like: TryFormatFromEndOfBuffer.
